src/layer/Layer.Popup.js

```diff
@@ -41,10 +41,39 @@ L.Layer.include({
 		return this;
 	},
 
-	openPopup: function (latlng) {
+	openPopup: function (target) {
+		var layer;
+		var latlng;
+
+		// handles figuring out `layer` and `latlng` from `target`
+		// assumes target will be one of
+		// * undefined
+		// * Layer
+		// * [lat,lng]
+		// * LatLng
+
+		if (!target) {
+			for (var id in this._layers) {
+				layer = this._layers[id];
+				break;
+			}
+			layer = layer || this;
+			latlng = layer._latlng || layer.getCenter();
+		} else if (target instanceof L.Layer) {
+			layer = target;
+			latlng = layer._latlng || layer.getCenter();
+		} else {
+			layer = this;
+			latlng = target;
+		}
+
 		if (this._popup && this._map) {
-			this._map.openPopup(this._popup, latlng || this._latlng || this.getCenter());
+			this._popup.options.offset = this._popupAnchor(layer); // update the popup offset based on our layer
+			this._popup._source = layer; // update popup source
+			this._popup.update(); // update the popup (will update content if popup uses a function)
```

Comment:
Any side effects to this for non-function popups compared to how it worked before?